cachi2/core/package_managers/rpm/redhat.py

```diff
@@ -0,0 +1,168 @@
+
+    arch: str
+    packages: list[Package] = []
+    source: list[Package] = []
+
+
+class Root(BaseModel):
+    """Top of the structure of the lockfile. Used for parsing the lockfile."""
+
+    lockfileVersion: PositiveInt
+    lockfileVendor: str
+    arches: list[Arch]
+
+
+class Header(BaseModel):
+    """
+    The class represents header of the lockfile. Used for matching the format.
+
+    Contains just a minimal set of fields to identify the format.
+    """
+
+    lockfileVersion: PositiveInt
+    lockfileVendor: str
+
+
+class RedhatRpmsLock(UserDict):
+    """
+    The class implements basic operations with specific lockfile format.
+
+    Input data comes from raw yaml stored as a dictionary.
+    """
+
+    # parsed structure of the lockfile
+    _lockfile: Root
+    # metadata of file records found in the lockfile (for verification and SBOM generation)
+    _files_metadata: dict[Path, Any] = {}
+    # 'repoid' key is not mandatory in the lockfile format. When not present,
+    # fallback is set as (partly) random string containing _uuid
+    _uuid = uuid.uuid4().hex[:6]
+
+    @property
+    def internal_repoid(self) -> str:
+        """Internal_repoid getter."""
+        return f"cachi2-{self._uuid}"
+
+    @property
+    def internal_source_repoid(self) -> str:
+        """Internal_source_repoid getter."""
+        return f"cachi2-{self._uuid}-source"
+
+    def is_internal_repoid(self, repoid: str) -> bool:
+        """Repoid directory matches the internal repoid."""
+        return repoid in (self.internal_repoid, self.internal_source_repoid)
+
+    @property
+    def metadata(self) -> dict[Path, Any]:
+        """Metadata getter."""
+        return self._files_metadata
+
+    def match_format(self) -> bool:
+        """Parse lockfile header with pydantic and evaluate whether header \
+        matches the format specification."""
+        try:
+            header = Header(**self.data)
+        except ValidationError:
+            return False
+        return header.lockfileVendor == "redhat" and header.lockfileVersion == 1
+
+    def process_format(self) -> bool:
+        """Parse the whole lockfile with pydantic and return whether \
+        the parsing was successful and format of the lockfile is correct."""
+        try:
+            self._lockfile = Root(**self.data)
+        except ValidationError as e:
+            log.error(str(e))  # log defail of the validation error
+            return False
+        return bool(self._lockfile)
+
+    def download(self, output_dir: Path) -> None:
+        """
+        Download packages mentioned in the lockfile.
+
+        Go through the parsed lockfile structure and find all RPM and SRPM files.
+        Create a metadata structure indexed by destination path used
+        for later verification (size, checksum) after download.
+        Prepare a list of files to be downloaded, and then download files.
+        """
+        for arch in self._lockfile.arches:
```

Comment:
Looping over the model kills the whole purpose of declaring this class as a `UserDict`, so this loop needs be over the RedhatRpmsLock dict `arches` subdict.